compiler/rustc_mir_transform/src/required_consts.rs

```diff
@@ -15,8 +15,18 @@ impl<'a, 'tcx> RequiredConstsVisitor<'a, 'tcx> {
 impl<'tcx> Visitor<'tcx> for RequiredConstsVisitor<'_, 'tcx> {
     fn visit_constant(&mut self, constant: &Constant<'tcx>, _: Location) {
         let literal = constant.literal;
-        if let Some(ct) = literal.const_for_ty() && let ConstKind::Unevaluated(_) = ct.kind() {
-            self.required_consts.push(*constant);
+        match literal {
+            ConstantKind::Ty(c) => match c.kind() {
+                ConstKind::Unevaluated(uv) => {
+                    let literal = ConstantKind::Unevaluated(uv.expand(), c.ty());
+                    let new_constant =
+                        Constant { span: constant.span, user_ty: constant.user_ty, literal };
+                    self.required_consts.push(new_constant);
+                }
+                _ => {}
```

Comment:
should be unreachable, the only `Ty` constants in unsubstituted mir should be `Param`.

Reply:
Why is that? We can't evaluate all constants early (https://rust-lang.zulipchat.com/#narrow/stream/260443-project-const-generics/topic/valtrees/near/287092100) during mir building, so where would we evaluate them between mir building and mir transform?

Reply:
Does mir transform happen after monomorphization?

Reply:
the only `ty::Const` which we start with in the mir is `ty::ConstKind::Param` unless we have some other subst (which we don't until inlining or when calling the function).

All constants which start as unevaluated, e.g. promoteds, inline consts and associated consts are all mir consts, not ty consts